### ISSUE TITLE
Documenation and Tooling: Constraining Architectures

### ISSUE THREAD

**Issue by srkunze** [lxc/lxd#8395]
Currently figuring out https://github.com/haproxy/haproxy/issues/1035

It could be an issue with the used ARM-Architecture but not for sure. @wtarreau saw the difference between the host and container architecture.

Reading the docs here https://lxd.readthedocs.io/en/latest/architectures/ it doesn't seem to be an issue but apparently I do not know.

As it would have be a no-brainer to use the arm64 version instead of the armhf version to rule out this bug possibility, it would have been great if LXD could complain during "lxc launch". And also a hint in the docs can be useful if different architectures are really an issue.

What do you think?

**Comment by stgraber** [comment]
So are you saying that `lxc launch ubuntu:20.04` got you an armhf container instead of arm64?

**Comment by stgraber** [comment]
Running armhf containers on arm64 host can indeed be a bit funny because of how the kernel reports this as `armv8l`.

The Ubuntu kernel has a custom patch which lets you set `compat_uts_machine=armv7l` on the kernel command line to override that name to `armv7l`, avoiding such issues. But that's very much an Ubuntu specific hack so something better suited for distro documentation than for LXD's.

**Comment by srkunze** [comment]
To be honest, not quite. I used

```
lxc launch ubuntu:20.04/armhf
```

So, now I realize that it might just work without the architecture spec but I still think it might be good idea to warn users of that they might try something "funny" here :-D

Don't remember where I copied it from but  back then it seems like as if the /armhf should be there.

**Comment by srkunze** [comment]
@stgraber Just to understand this bit here correctly because I am not familiar with it: you are saying that haproxy should be able to handle armhf regardless of the underlying hardware, right?

**Comment by stgraber** [comment]
Yeah, if all that makes it crash is seeing `armv8l` from the kernel rather than `armv7l`, then that definitely seems like a bug in the application.

If that's not the issue, then the problem could be that haproxy was built with some hardware extensions enabled which somehow aren't present on the 32bit API of the raspberry pi. This certainly can happen and isn't limited to Arm either. You can get into the same kind of problem on Intel by building with a compiler that uses the most recent CPU extensions and then attempt to run the binary on an older CPU. Everything will behave up until the time where the affected instruction is called, at which point the program will crash.

**Comment by wtarreau** [comment]
@stgraber indeed, it took us some time to understand that it was a 32-bit binary running on a 64-bit host, and I agree with you that normally it should work flawlessly. What puzzles me is that the user receives a SIGBUS making me think about something like an alignment issue, while this is the regular binary from the package used by default on native ARM machines (and I could double-check that it does work fine on armv8 with the compat libs). Initially I thought he was building it himself and that gcc there could have advertised some __ARM_foo features that wouldn't work fine, but given that it's the standard armhf package, I'm a bit puzzled for now.
 I'm wondering if it's possible that *something* differs there. The user does nothing fancy and (except for openssl which doesn't seem to be directly related) the single request he's doing doesn't pass through any fancy lib. Strangely he cannot get any core, nor get gdb to work from this environment, and doesn't even have any kernel message announcing the death cause, so that obviously doesn't help.

I know we do have some optimizations for unaligned data that are enabled when the  platform supports them (and this one does advertise them). For now we don't know more. We're still investigating. I'm fine with passing you more info once we figure the cause. Conversely any debugging hint from your about this environment is welcome :-)  I'm trying to setup one for the first time on a machine here, I'll see if I manage to do it right.

**Comment by stgraber** [comment]
@wtarreau @srkunze do you have a pretty short reproducer for the issue?

I have access to rpi4 and a variety of other arm64 platforms so would be interesting if I could reproduce it on rpi4 and then see if the same happens on APM/Ampere silicon. I'm always a bit wary of the rpi because of odd architectural limitation in the past, though I must say that the rpi4 seems like a pretty solid aarch64 platform (minus it still using u-boot...)

**Comment by wtarreau** [comment]
So the first visible thing in his image is that just issuing `gdb --args /bin/id` then `run` loops forever inside ld-linux.so. So I conclude that at this point there's something fishy in this setup, even though I have no idea whether it's related to the combination of archs or just a software bug somewhere. I'll let @srkunze provide all the details to avoid saying stupidities.

And I agree with you that rpi always looks suspicious :-) With this said, the rpi4 uses regular A72 cores and usually works without surprises (at least with a correct power supply and micro-SD).

**Comment by srkunze** [comment]
Here we go:

```
- buy a raspi 
- install ubuntu 20.04 arm64
```

```
root@raspi:~# apt install lxd
root@raspi:~# lxc launch ubuntu:20.04/armhf haproxy

root@raspi:~# lxc shell haproxy
root@haproxy:~# apt install gdb
```


Checks for the setup
```
root@raspi:~# uname -a
Linux raspi 5.4.0-1026-raspi #29-Ubuntu SMP PREEMPT Mon Dec 14 17:01:16 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux

root@raspi:~# lxc shell haproxy
root@haproxy:~# uname -a
Linux haproxy 5.4.0-1026-raspi #29-Ubuntu SMP PREEMPT Mon Dec 14 17:01:16 UTC 2020 armv8l armv8l armv8l GNU/Linux

```

**Comment by wtarreau** [comment]
@stgraber I finally nailed the crash issue down. It's interesting. Some code we embed (xxhash) is built to support unaligned accesses when `__ARM_FEATURE_UNALIGNED` is set. But on armv5-7, 64-bit unaligned accesses with the LDRD instruction are not permitted, however they are emulated by the kernel.

But the arm64 kernels do not implement the alignment emulation since armv8 processors surpport all types of unaligned accesses. However when they execute code in compatibility mode, they still cause an alignment exception, without any code in the kernel to catch them. As such it could be argued that the real difference is the kernel, and that there's no 100% guarantee that a 32-bit program will run fine on a 64-bit kernel.

I've proposed a patch to xxhash to use a separate condition for 64-bit unaligned accesses.

Thus there's no issue with LXC but next time you get a similar report you might have one more clue where to look :-)

**Comment by stgraber** [comment]
Thanks for getting to the bottom of this!

I wonder if it would be worth a kernel patch to catch such unaligned access from the 32bit personality and emulate the normal Linux behavior for 32bit arm. That would take care of such issues and quite possibly of some other weird issues folks have been running into :)

**Comment by wtarreau** [comment]
Yes I think it would be worth doing, but I predict that it could be a bit controversial to add support for alignment traps on a platform that theoriticaly always supports unaligned accesses. Also I remember there was not that much love for this feature for ARMv5 as there, **by default**, unaligned accesses would simply return crap instead of being emulated. You had to manually enable alignment traps at boot.

**Comment by srkunze** [comment]
Also big kudos from my side! +1

Just one noob question: is this something that will sort itself out in time? If not, I would also favor a kernel solution here.

From a user's perspective, it is massively frustrating (I can tell you that) and I was lucky that @wtarreau was so patient with me to get me to that point where we found out that the gdb is broken on armhf-on-arm64 systems. I usually trust these tools and do not question their functional integrity.

**Comment by wtarreau** [comment]
The unaligned access issue is almost addressed. I'm going to merge the patch into haproxy's copy of xxhash and passed it to Yann already for xxhash. It's uncertain whether latest versions need it or not, as this optimization is sensitive to the code structure and compiler version (I couldn't reproduce it on more recent versions nor with an older compiler).

The gdb issue is another one and could take more time as it needs to be studied by whoever knows this kernel area best.

Regarding your trust of tools, the important rule is never to trust anything, starting with your own code. Bugs are everywhere and everything not completely observable is suspicious. You can only reason in probabilities to help narrow an issue faster, but no observed proof means no certainty nor direct conclusion.

**Comment by srkunze** [comment]
> The gdb issue is another one

I researched the gdb bug database and found: https://sourceware.org/bugzilla/show_bug.cgi?id=25215

Do you think it's the same?

**Comment by wtarreau** [comment]
It's possibly the same, yes.
